Add LSP functionality to Rover

jonathanrainer opened the pull request:
Now that the LSP has been finalised, and composition has been refactored such that it can consume events in the same way that `rover dev` does we can unify the two together. 

This has been tested by running a custom build version of Rover against VSCode and it seems to function as intended. We will need to hash out a few more things in the review overall but this is a very positive step forwards

jonathanrainer committed:
ROVER-245 Fix up Cargo.tomls and deny.toml

jonathanrainer committed:
ROVER-245 Fix clippies

jonathanrainer committed:
ROVER-245 Integrate existing LSP code

Takes the approach of dumping the existing LSP
code straight into what we have and integrating
it to the point of compilation.

jonathanrainer committed:
ROVER-245 Removing deprecated PanicInfo and other fixes

Clumps together other minor fixes without them
dirtying other bits of the commit history.

jonathanrainer commented on crates/robot-panic/src/lib.rs:
The work here is clearing up some stuff that should have been a while back after we upgraded to a new version of Rust. I think this hasn't been picked up because often we don't do release builds, which I had to do as part of this

jonathanrainer commented on crates/rover-std/Cargo.toml:
This is just clearing up some unused dependencies, nothing to see here really

jonathanrainer commented on src/command/lsp/README.md:
Not sure if we still need this file or if we just want to remove it now? I've left it here as it was part of the original PR but can easily remove it if not

jonathanrainer commented on src/command/lsp/input.txt:
Same comment with this file, we can easily remove it if it's no longer necessary

jonathanrainer commented on src/command/lsp/mod.rs:
This is where the real meat of the PR is, as this is where we wire together our work on Composition with the LSP itself

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+
+async fn run_lsp(client_config: StudioClientConfig, lsp_opts: LspOpts) -> RoverResult<()> {
+    //TODO: Check this error handling is right i.e. if we don't get a supergraph.yaml passed
+    // this should fail rather than doing something else.
```
This one I wasn't particularly sure about, in my view, with this command, we should make passing a Supergraph YAML compulsory because I can't really understand what not passing one would even mean. That would mean we could make `clap` take it as a Path, rather than an Option and tidy this code up a bit.

aaronArinder replied:
also not sure; I can't tell from the crate docs whether the lsp is meant to be only for supergraphs (ie, federation) or if there's some other use for it

I would have assumed only supergraphs, but there's an odd option in the config for the lsp: [force federation](https://docs.rs/apollo-language-server/0.3.0/apollo_language_server/struct.Config.html#structfield.force_federation), which forces all graphs to be treated as subgraphs. Maybe that means you can use it on monographs? Do they have a supergraph.yml or are they complete as they are? If they still require a supergraph.yml, I think making it compulsory makes sense

dylan-apollo replied:
The language server does standard GraphQL validations all on its own, outside of federation. So you can totally use our extension with `rover lsp` for a monograph if you don't pass in a `supergraph.yaml` file. Passing in that file is what triggers the composition mode—so most of this logic. If you don't pass in the file, `rover lsp` is mostly just a passthrough to the language server crate.

jonathanrainer replied:
@dylan-apollo In that case, suppose we don't get a `supergraph.yaml` passed, what value(s) should be sent to the `ApolloLanguageServer::build_service` function? 

How do you know what graph you're supposed to be targetting?

dylan-apollo replied:
It should already work as-is, the code on my branch collected into an empty hashmap for the subgraphs and then just didn't spawn the composition worker.

> How do you know what graph you're supposed to be targetting?

I don't understand this question

jonathanrainer replied:
Sorry, that's me being dim, I apologise, I've made the change now so that if we don't get a `supergraph.yaml` we bail out early before doing any composition etc. and then everything else can continue as is.

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+        let federation_version = lazily_resolved_supergraph_config
+            .federation_version()
+            .clone()
+            .unwrap_or(FederationVersion::LatestFedTwo);
```
Have we centralised the behaviour on detecting the Federation Version? I feel like we definitely should if we haven't already

aaronArinder replied:
We've a rough outline of it [here](https://apollographql.atlassian.net/wiki/spaces/Rover/pages/813597381/2024-08-19+Refactoring+rover+dev#Federation-Version-Resolution), but I think only the `FullyResolvedSupergraphConfig` captures that via its `resolve_federation_version` method. I could be wrong about that, though

Here's how I think the federation version is being resolved in this file, though:

- default SupergraphConfigResolver starts the federation version at `None`
- remote loading of subgraphs doesn't change the federation version, so it's still at `None`
- loading the supergraph config yaml from file will take the federation version from that file if it exists, but otherwise will keep it the same (`None`); so, now we either have a config-defined fed version or just `None`
- lazily resolving the subgraph config will take either the config-file defined fed version or `None`
- finally, you're defaulting to fed 2

so, either it comes from the config file or it gets defaulted to fed 2, which seems alright, but we aren't _really_ respecting the order in the confluence link above (or centralizing it in a way that is helpful rather than confusing/harmful)

aaronArinder replied:
oh, came across this in the source while looking at something else: [panic when calling composition_did_update on a non-supergraph](https://docs.rs/apollo-language-server/0.3.0/src/apollo_language_server/server.rs.html#411), so maybe requiring the supergraph config yaml is the strat!

this also raises a question about how panics are being handled; what happens in rover when the LSP server takes a dirt nap from some internal panic?

aaronArinder replied:
for posterity: I believe @dotdat is working on this?

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+                    lsp_opts.plugin_opts.elv2_license_accepter,
+                    lsp_opts.plugin_opts.skip_update,
+                )
+                .await?;
```
One of the features (I think) of the old PR is that it allows people to update the version of Federation inside the `supergraph.yaml` while it's running. When doing the `dev` refactor we made an explicit decision that we didn't want to support that behaviour. Is that a dealbreaker for the LSP? Or could we ship this as is and see if there's sufficient demand after the fact to implement that?

dylan-apollo replied:
It's a pretty bad UX to not watch `supergraph.yaml` because the way to manually restart the language server underneath VS Code is really clunky—and it's not at all obvious when you should do that. If someone modifies their `supergraph.yaml` file and still sees out of date validations, it's going to be frustrating.

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+            .setup_supergraph_config_watcher(lazily_resolved_supergraph_config.clone())
+            .setup_composition_watcher(
+                lazily_resolved_supergraph_config
+                    .fully_resolve_subgraphs(&client_config, &studio_client)
```
This method is new because in the original version of the refactor you either had a FullyResolvedSupergraphConfig (so all the elements had been reduced to SDL sources), or a LazilyResolvedSupergraphConfig (which still had the capacity to spin up FileWatchers etc.). 

The issue in this instance is that you need both (or specifically you need a way to obtain `FullyResolvedSubgraphs`). So I decided that it was best to implement a method to convert a LazilyResolvedSupergraphConfig to a FullyResolvedSubgraphs. Because this transformation is only possible one way as you lose information once they're fully resolved.

Happy to argue this one out, because it is a bit weird that the Runner requires a LazilyResolvedSupergraphConfig at the top and then FullyResolved in later methods but didn't want to touch that code too much due to the ongoing `dev` refactor

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+                        )
+                        .await;
+                }
+                CompositionEvent::Error(CompositionError::Build { source: errors }) => {
```
This is slightly different to the original version because the `dev` refactor does not distinguish at the top level between "Completing with errors" and "Erroring". So we have to pattern match on the error itself to get what we want.

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+                CompositionEvent::Error(err) => {
+                    debug!("Composition {federation_version} failed: {err}");
+                    // TODO: we could highlight the version of federation, since it failed.
+                    let message = format!("Failed run composition {federation_version}: {err}",);
```
Not sure what this comment means, it came over verbatim from the original PR

dylan-apollo replied:
Basically instead of using `Range::default()` below, which results in no red squigglies in any files, if we start retaining location data when we parse `supergraph.yaml`, we could highlight part of that file when this error happens. The version is the most likely place, since it means there was something wrong with running the composition binary, not something wrong with actually composing.

jonathanrainer replied:
Ah! I get what you mean, do we just want to leave this as a TODO then? Or do you want to make this change as we integrate?

dylan-apollo replied:
I don't think it's critical, I had it as a TODO in case I got time to improve it but.. clearly I didn't 😛. I think it'd require changing `apollo-federation-types` to use a parsing library which retains location data (instead of `serde_yaml` or whatever), so probably more of a change than we want unless a customer pushes for it.

In short, just delete the TODO is probably fine

jonathanrainer commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![diagnostic])
+                        .await;
+                }
+                CompositionEvent::SubgraphAdded(CompositionSubgraphAdded {
+                    name,
+                    schema_source,
+                }) => {
+                    debug!("Subgraph {} added", name);
+                    language_server.add_subgraph(name, schema_source).await;
+                }
+                CompositionEvent::SubgraphRemoved(CompositionSubgraphRemoved { name }) => {
+                    debug!("Subgraph {} removed", name);
+                    language_server.remove_subgraph(&name).await;
+                }
```
These are net new events because the language server needs to know when subgraphs are added and removed, so I expanded out the definition of CompositionEvents to include these.

aaronArinder replied:
trying to get an initial review out, but self-note: probs need to check that all the places we add/remove subgraphs are emitting the right event

jonathanrainer commented on src/composition/mod.rs:
```diff
@@ -19,9 +20,9 @@ mod watchers;
-    supergraph_sdl: String,
-    hints: Vec<BuildHint>,
-    federation_version: FederationVersion,
+    pub(crate) supergraph_sdl: String,
+    pub(crate) hints: Vec<BuildHint>,
+    pub(crate) federation_version: FederationVersion,
```
`pub(crate)` might not be right here, they could probably all be public since they're events

dotdat replied:
I think for this, `pub` seems better and since they end up being public, we can remove the `derive(Getters)` above and just treat this object as pure data.

jonathanrainer replied:
Good thought, have done :)

jonathanrainer commented on src/composition/supergraph/config/resolve/mod.rs:
```diff
@@ -193,6 +192,39 @@ impl LazilyResolvedSupergraphConfig {
+        } else {
+            Err(errors)
+        }
+    }
```
This is very similar to the method for doing this for the FullyResolvedSupergraphConfig but with enough difference to not want to refactor it into a single method

dotdat replied:
I wonder if we want to look at the two, largely because resolving the federation version consistently would be useful. The current implementation here doesn't really take into account subgraphs, unless I missed something.

jonathanrainer commented on src/composition/supergraph/config/resolve/mod.rs:
```diff
@@ -227,8 +259,8 @@ impl FullyResolvedSubgraphs {
-    pub fn upsert_subgraph(&mut self, name: String, schema: String) {
-        self.subgraphs.insert(name, schema);
+    pub fn upsert_subgraph(&mut self, name: String, schema: String) -> bool {
+        self.subgraphs.insert(name, schema).is_none()
```
Slightly changing the upsert mechanism allows us to see if we're adding a subgraph for the first time so we can emit the right events for the LSP

aaronArinder replied:
don't _love_ this return type because I'm not sure what it means (outside of this comment); worth calling it out in its comment why it returns a bool? or, how do you feel about a more self-documenting enum whose variants capture what we'd want to know? (or, maybe better: should we just move the logic for telling whether we're adding subgraphs for the first time to the bit of code that actually cares about it? that'd better separate what we're up to)

jonathanrainer replied:
Another option might be to just rename this as `insert_subgraph` and pass back the return type from the `insert` method? That's probably more idiomatic

aaronArinder replied:
that sounds good!

jonathanrainer commented on src/composition/supergraph/config/resolve/mod.rs:
```diff
@@ -267,7 +299,10 @@ impl From<FullyResolvedSubgraphs> for SupergraphConfig {
+                    // This value is effectively unused but has to be set to something that
+                    // is not "null" when serialised otherwise the supergraph binary
+                    // complains
+                    routing_url: Some(String::from("localhost")),
```
This was a real surprise and not something I'd considered before, but without this composition did not work. We may want a different nonsense value (would the empty string work?)

aaronArinder replied:
huh, yeah, weird; maybe an empty string would be good? or, some other weird value like "NOOP VALUE" or something to capture that we're the ones who've added it and that we don't expect it to be useful (apart from getting composition to pass); I could see myself being pretty confused by the localhost default later on if I were debugging something with composition

dylan-apollo replied:
Where is this conversion used? The routing URLs _are_ important because that's how the router knows where to send subgraph traffic.

jonathanrainer replied:
Yer you're right, I think we're losing too much information when we make our FullyResolvedSubgraphs object, I'll get that changed.

jonathanrainer commented on src/composition/supergraph/config/resolve/subgraph.rs:
```diff
@@ -1,32 +1,35 @@
         /// Supplied path to the subgraph schema file
         path: PathBuf,
+        /// The result of joining the paths together, that caused the failure
+        joined_path: PathBuf,
```
Found the need to expand the definition of this error to aid my own debugging

jonathanrainer commented on src/composition/supergraph/config/resolve/subgraph.rs:
```diff
@@ -138,78 +142,153 @@ impl FullyResolvedSubgraph {
+    }
+
+    /// Fully resolves a [`LazilyResolvedSubgraph`] to a [`FullyResolvedSubgraph`]
+    pub async fn fully_resolve(
```
Resolving and fully resolving are quite similar so I pulled out the common parts below and left us with a structure that is similar but allows the relevant differences to be drawn out

aaronArinder replied:
our language around resolution is super confusing; `resolve` and `fully_resolve` both return a `FullyResolvedSubgraph`, but the latter makes it sound like "this time I'm serious about fully resolving it" rather than the inputs being different

not sure there's anything to do about it, but it's confusing

aaronArinder replied:
the way github interleaves the diff makes it hard for me to really understand the differences between the two; probably worth having @dotdat take a stab at it

jonathanrainer commented on src/composition/watchers/composition.rs:
This part is simply augmenting the composition system to emit the right events now that we care about subgraph additions and removals.

jonathanrainer replied:
It also fixes a test that this broke

jonathanrainer commented on src/utils/supergraph_config.rs:
These were some leftover clippies that I fixed to get CI passing

aaronArinder commented on src/cli.rs:
```diff
@@ -432,6 +434,11 @@ pub enum Command {
+
+    /// Start the language server
+    #[cfg(feature = "composition-js")]
+    #[clap(hide = true)]
```
to doublecheck, we want this hidden?

dylan-apollo replied:
Yes, this is supposed to be an implementation detail of the VS Code extension right now, not something we broadly support (i.e., no public/stable API requirement). Maybe some day we'll lock it down, but probably not until we're done tweaking it for our _own_ extensions.

aaronArinder commented on src/command/lsp/README.md:
```diff
@@ -0,0 +1,14 @@
+# Build / configuration
+
+1. Clone mdg-private/language-server repo (adjacent to this repo, the language
```
this is sort of weird; we should hide the mdg-private callout, but also figure out what this readme is even supposed to convey (and who it's being conveyed to)

dylan-apollo replied:
It was for us, from earlier versions before the language server crate was published. This can be removed now

aaronArinder commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+            }
+        })
+        .ok_or_else(|| anyhow!("Could not find supergraph.yaml file."))?;
+    let supergraph_content_root = supergraph_yaml_path.parent().unwrap().to_path_buf();
```
kill the unwrap?

aaronArinder commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+            enable_auto_composition: false,
+            force_federation: false,
+            disable_telemetry: false,
+        },
```
I wonder if we should do something else with these options. Eg, the disable_telemetry should probably follow `APOLLO_TELEMETRY_DISABLED`, but I don't really understand `force_federation`'s use or when you'd want to enable or diasble auto-composition (I guess if you're not relying on rover's composition pipeline?)

aaronArinder commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,230 @@
+            match event {
+                CompositionEvent::Started => {
+                    language_server.composition_did_start().await;
+                }
```
there's some interesting stuff the LSP is doing; this sends a `WorkDoneProgressBegin` event that has a corresponding `WorkDoneProgressEnd` event that gets emitted from `composition_did_update`. I feel like that's some important context worth calling out: rover _must_ always send that `WorkDoneProgressBegin` event before the `WorkDoneProgressEnd` event (according to the [LSP docs](https://docs.rs/apollo-language-server/0.3.0/apollo_language_server/struct.ApolloLanguageServer.html)), but I'm not really sure how strong of a 'must' that is

[here's](https://docs.rs/apollo-language-server/0.3.0/src/apollo_language_server/server.rs.html#406) where that method bottoms out, insofar as I can tell. It doesn't look especially dangerous to not emit the right event before emitting the `WorkDoneProgressEnd` event, but there's probably something in the LSP that would get upset about it; either way, might be worth calling out that there's a required-yet-implicit ordering so that future devs don't think we can ignore `CompositionEvent::Started` or something silly like that

aaronArinder commented on src/composition/supergraph/config/resolve/mod.rs:
```diff
@@ -193,6 +192,39 @@ impl LazilyResolvedSupergraphConfig {
+
+    /// Fully resolves a [`LazilyResolvedSupergraphConfig`] into a [`FullyResolvedSubgraphs`]
+    /// by retrieving all the schemas as strings
+    pub async fn fully_resolve_subgraphs(
```
worth calling out that this is a one-way conversion because information gets lost? Might help a future dev know whether this is the right thing to reach for

aaronArinder commented on src/composition/supergraph/config/resolve/subgraph.rs:
```diff
@@ -1,32 +1,35 @@
     /// supergraph config file
-    #[error("Could not find schema file ({path}) relative to ({supergraph_config_path}) for subgraph `{subgraph_name}`")]
+    #[error("Could not find schema file ({path}) relative to ({supergraph_config_path}) for subgraph `{subgraph_name}`"
+    )]
```
weird that this passes linting, but who am I to defy clippy

aaronArinder reviewed (approved):
slapping a like & subscribe on this bad boy because I think it's mergeable as-is, though I know you'll make certain changes here and there (request re-review if you want); I haven't actually tried this out in the wild, though, but not sure that's really on our team to do

dotdat commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,289 @@
+use futures::StreamExt;
+mod errors;
+
+use crate::command::lsp::errors::SupergraphConfigLazyResolutionError;
```
Totally a nit, but I think in most other places we have an imports grouped by:
1. std
2. 3rd party
3. crate
4. super

jonathanrainer replied:
Yer, my Rustfmt was being weird, not sure what happened there. It's fixed now :)

dotdat commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,289 @@
+        }
+    });
+
+    // Early return if there is no `supergraph.yaml` given as there is no further need to construct
```
I'm a little confused about the early return comment here. Did something change?

jonathanrainer replied:
Not really, I'd misunderstood something in the original semantics that Dylan had implemented so this restores it. Basically if a user doesn't supply a `supergraph.yaml` we still want to start the LSP up but we don't need the ability to recompose so there's no need to start the composition watcher stuff. It was implemented a bit differently previously but the semantics are now the same. Have a look at Dylan's PR if you want to compare/contrast.

dotdat commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,289 @@
+                    lazily_resolved_supergraph_config
+                        .subgraphs()
+                        .iter()
+                        .map(|(a, b)| (a.to_string(), b.schema().clone())),
```
nit: would `.cloned()` work here instead of `map`?

jonathanrainer replied:
Had a fiddle around and I don't think so unless I'm missing something obvious, if you want to craft something then feel free to have a go :)

aaronArinder reviewed (approved):
seems good to me!

jonathanrainer committed:
ROVER-245 Add code to handle adding/removing subgraphs

We need to emit extra events here such that the LSP can react
to them

jonathanrainer committed:
ROVER-245 Upgrade to language-server 0.3.4

jonathanrainer committed:
ROVER-245 Add new InMemory OutputTarget

At present, we only support 2 composition output targets, this
adds a third, which means that the output of composition only
exists in memory. This means we don't have to write/manage temporary
files, and we don't compete with the LSP for usage of stdout.

jonathanrainer committed:
ROVER-245 Rebasing after Christmas break

dylan-apollo commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,346 @@
+        None => {
+            let (service, socket, _receiver) = ApolloLanguageServer::build_service(
+                Config {
+                    // TODO Do we need to worry about these now?
```
@Jephuff can you double-check that this seems right?

dylan-apollo replied:
(and also the similar block below)

Jephuff replied:
These all look good to me 👍

dylan-apollo commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,346 @@
+                }
+                CompositionEvent::Error(err) => {
+                    debug!("Composition failed: {err}");
+                    let message = format!("Failed run composition: {err}",);
```
```suggestion
                    let message = format!("Failed run composition: {err}");
```

WTB rustfmt which has an opinion on popular macros.

dylan-apollo reviewed (approved):
Looks good to me (ignoring the couple tickets we're tracking separately)! I think we're basically at the point where we need to start using it regularly to find any edge cases. Thanks all for powering through this epic 🫶

jonathanrainer committed:
ROVER-245 Add polling_interval option

We want to set a sensible default here (5 seconds), while giving
users the ability to override it if that becomes problematic.

jonathanrainer committed:
ROVER-245 Remove TODOs and fix up

aaronArinder commented on Cargo.toml:
```diff
@@ -73,6 +73,8 @@ apollo-encoder = "0.8"
 # https://github.com/apollographql/federation-rs
 apollo-federation-types = "0.14.1"
 
+apollo-language-server = { version = "0.3.4", default-features = false, features = ["tokio"] }
```
is the latest version? I want to say I saw a bump recently; might be worth a follow-up with it, if there was one

dylan-apollo replied:
[0.3.5](https://crates.io/crates/apollo-language-server/0.3.5)

aaronArinder reviewed (approved):
lgtm!
